docs/proposals/20220210-md-node-label-sync.md

```diff
@@ -0,0 +1,161 @@
+
+```
+node-role.kubernetes.io/*
+node-restriction.kubernetes.io/*
+```
+
+The `node-role.kubernetes.io` prefix has been used widely in the past to identify the role of a node (e.g. `node-role.kubernetes.io/worker=''`). For example, `kubectl get node` looks for this specific label when displaying the role to the user.
+
+The use of the `node-restriction.kubernetes.io/*` prefix is recommended in the [Kubernetes docs](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction) for things such as workload placement.
+
+It's important to note that the these specific labels are *not* kept in sync, but rather applied only once after the Machine and its corresponding Node are first created. Taking authoritative ownership of these prefixes would preclude other entities from adding and removing such labels on the Node. 
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin/user, I would like a declarative and secure means by which to assign roles to my nodes. For example, when I do `kubectl get nodes`, I want the ROLES column to list my assigned roles for the nodes.
+
+#### Story 2
+
+As a cluster admin/user, for the purpose of workload placement, I would like a declarative and secure means by which I can add/remove labels on groups of nodes. For example, I want to run my outward facing nginx pods on a specific set of nodes (e.g. md-secure-zone) to reduce exposure to the company's publicly trusted server certificates.
+
+### Implementation Details/Notes/Constraints
+
+#### Label propogation from MachineDeployment to Machine
+
+Utilize CAPI metadata Propagation. CAPI supports [metadata propagation](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/metadata-propagation.html?highlight=propagation#metadata-propagation) across core API resources. For the MachineDeployment types, template labels propagate to MachineSets top-level and MachineSets template metadata:
+```
+MachineDeployment.spec.template.metadata.labels => MachineSet.labels, MachineSet.spec.template.metadata.labels
+```
+Similarly, on MachineSet, template labels propagate to Machines:
+```
+MachineSet.spec.template.metadata.labels => Machine.labels
+```
+
+As of this writing, changing the `MachineDeployment.spec.template.metadata.labels` will trigger a rollout. This is undesirable. CAPI will be updated to instead ignore updates of labels that fall within the CAPI prefix. There is precedence for this with the handling of `MachineDeploymentUniqueLabel`. 
+
+#### Synchronization of CAPI Labels
+
+Labels that fall under the CAPI specific prefix `node.cluster.k8s.io/*` will need to be kept in sync between the Machine and Node objects. Synchronization must handle these two scenarios: 
+(A) Label is added/removed on a Machine and must be added/removed on the corresponding Node.
+(B) Label is added/removed on the Node and must be removed/re-added on the Node to bring it in sync with the labels on Machine.
+
+
+#### Machine Controller Changes
```

Comment:
Given that this proposal is implemented at Machine controller level, I assume that the gap for extending this to control plane machines is minimal. Can we reconsider including this in the proposal scope?

Reply:
+1 I think it's easier to implement this feature for all Machines (CP & MachinePool) vs. implementing something that it only works for MachineSet/MachineDeployment managed machines.

Reply:
More than happy to expand this to CP, but I wonder if we can first get this implemented for workers and work out any kinks along the way. The proposal can then be easily expanded to include CP.

Reply:
I think it's the other way around. If you want to only sync labels from "MachineDeployment Machines" you have to implement handling in the machine controller to figure out if the current Machine is managed by a MD/MS, KCP, MachinePool, some third-party tool (@enxebre :)).

I think it's easier to avoid adding something like that to the machine controller. IIRC the only MD specific change is that we have to avoid triggering rollouts. I think it might be easier to figure out how we make this happen for all "owners" vs how we make it happen that the feature only works for "MD Machines".

Reply:
I think let's for now figure this out how exactly we do this for MD's (https://github.com/kubernetes-sigs/cluster-api/pull/6255#discussion_r831812895) and then we can give a hint how to phrase this for KCP and MachinePools.

I think it comes down to:
1. Implement label sync in the MachinePool controller
2. Implement in-place label propagation MD=>MS=>Machine
3. Implement in-place label propagation KCP=>Machine
4. Implement in-place label propagation MachinePool=>Machine

We don't have to implement 3+4 right away. But I would prefer implementing them as follow-ups to the first implementation vs. having to implement something to block label sync for KCP/MachinePool machines after 1.

Reply:
> Did you mean Machine controller in 1?

Yes, sorry my bad. edited

Reply:
@Arvinderpal what do you think?

Reply:
We'll probably need the first 3 implemented right away to make sure there is basic functionality for this feature, otherwise it doesn't seem most folks would be able to use it?

Reply:
> We'll probably need the first 3 implemented right away to make sure there is basic functionality for this feature, otherwise it doesn't seem most folks would be able to use it?

Depending on complexity it might be a good idea to have 3 separate PRs to implement 1,2,3. The feature becomes really useful after 2,3. But after 1 it's already possible to propagate labels to nodes, just not to update them without a rolling update.

My main point is, I would really like to avoid to implement something in the Machine controller which enables/disables the label sync based on if the Machine is owned by a MachineSet/MachineDeployment or KCP.

Reply:
Yes, that sounds good to me. I'll update the proposal to incorporate KCP.

Reply:
Given the general agreement on this point, IMO updating the proposal for KCP support is a prerequisite before starting lazy consensus

Reply:
I have added the following to the goals:
```
- Support label synchronization for MachineDeployments as well as KubeadmControlPlane (KCP).
```

As noted above by @sbueringer we can start with MDs to iron out the details and incorporate the same functionality into KCP in follow up PRs.